test/jdk/java/nio/jni/libNewDirectByteBuffer.c

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+#include <stdlib.h>
+#include "jni.h"
+
+// private static native ByteBuffer newDirectByteBuffer(long addr, long size)
+JNIEXPORT jobject JNICALL
+Java_NewDirectByteBuffer_newDirectByteBuffer
+    (JNIEnv *env, jclass cls, jlong addr, jlong size)
+{
+    // Create the direct byte buffer, freeing the native memory if an exception
+    // is thrown while constructing the buffer
+    return (*env)->NewDirectByteBuffer(env, (void*)addr, size);
+}
+
+// private static native long getDirectByteBufferCapacity(ByteBuffer buf)
+JNIEXPORT jlong JNICALL
+Java_NewDirectByteBuffer_getDirectByteBufferCapacity
+    (JNIEnv *env, jclass cls, jobject buf)
+{
+    return (*env)->GetDirectBufferCapacity(env, buf);
+}
+
+// private static native long getDirectBufferAddress(ByteBuffer buf)
+JNIEXPORT jlong JNICALL
+Java_NewDirectByteBuffer_getDirectBufferAddress
+    (JNIEnv *env, jclass cls, jobject buf)
+{
+    return (jlong)(*env)->GetDirectBufferAddress(env, buf);
```

Comment:
The cast to jlong should be okay for 64-bit. For 32-bit then it will likely need a double cast, as in ((jlong)(int)( .. )) but maybe okay for now as you've restricted the test to 64-bit.

Reply:
Leaving it as-is for now.

Reply:
FTR this was the wrong choice as the test is still built on 32-bit (and fails to compile).